lib/ClangImporter/ImportDecl.cpp

```diff
@@ -1974,6 +1975,59 @@ applyPropertyOwnership(VarDecl *prop,
   }
 }
 
+/// Turn T into Unmanaged<T>, or into Unmanaged<T>? if `isOptional` is true.
+static Type getUnmanagedType(swift::ASTContext& Ctx,
+                             Type payloadType, bool isOptional) {
+  NominalTypeDecl *unmanagedDecl = Ctx.getUnmanagedDecl();
+  if (!unmanagedDecl || unmanagedDecl->getGenericParams()->size() != 1)
+    return payloadType;
+
+  Type result = BoundGenericType::get(unmanagedDecl, /*parent*/ Type(),
+                                      payloadType);
+  if (isOptional)
+    result = OptionalType::get(result);
+  return result;
+}
+
+static void
+applyPropertyOwnership(VarDecl *prop,
+                       clang::Qualifiers::ObjCLifetime lifetime) {
+  Type ty = prop->getInterfaceType();
+  bool wasOptional = false;
+  if (auto innerTy = ty->getOptionalObjectType()) {
+    ty = innerTy;
+    wasOptional = true;
+  }
+  if (!ty->is<GenericTypeParamType>() && !ty->isAnyClassReferenceType())
+    return;
+
+  ASTContext &ctx = prop->getASTContext();
+  if (lifetime == clang::Qualifiers::OCL_Weak) {
+    prop->getAttrs().add(new (ctx)
+                             ReferenceOwnershipAttr(ReferenceOwnership::Weak));
+    prop->setType(WeakStorageType::get(prop->getType(), ctx));
+    prop->setInterfaceType(WeakStorageType::get(
+        prop->getInterfaceType(), ctx));
+    return;
+  }
+  if (lifetime == clang::Qualifiers::OCL_ExplicitNone) {
+    // For __unsafe_unretained, we keep the old behavior of wrapping the type
+    // in Unmanaged instead of using UnmanagedStorageType. We need to make sure
+    // to pop the optionality out, though; e.g., `Foo?` becomes
+    // `Unmanaged<Foo>?`, not `Unmanaged<Foo?>`.
+    //
+    // TODO: Should we use UnmanagedStorageType? Unmanaged does feel safer
+    // because it makes the unowned-ness explicit in the type of the field,
+    // which is more obvious than `unowned(unsafe)` when imported from C.
+    prop->setType(getUnmanagedType(prop->getASTContext(), ty, wasOptional));
+    prop->setInterfaceType(getUnmanagedType(prop->getASTContext(), ty,
+                                            wasOptional));
+    return;
+  }
+  // OCL_None and OCL_Strong cause no change; the storage-less type is correct
+  // here. OCL_Autoreleasing is not supported by struct fields.
```

Comment:
If the unsupported one comes up you should still fail to import the struct (and test that it won't import)

Reply:
In this case, ClangImporter should emit an error preventing us from getting to this point: `error: fields cannot have __autoreleasing ownership`. But I'll throw in an `llvm_unreachable` so that we acknowledge it.